java/flight/flight-grpc/src/main/java/org/apache/arrow/flight/FlightGrpcUtils.java

```diff
@@ -17,19 +17,117 @@
 
 package org.apache.arrow.flight;
 
+import java.util.ArrayList;
 import java.util.Collections;
+import java.util.List;
 import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
 
 import org.apache.arrow.flight.auth.ServerAuthHandler;
 import org.apache.arrow.memory.BufferAllocator;
 
 import io.grpc.BindableService;
+import io.grpc.CallOptions;
+import io.grpc.ClientCall;
+import io.grpc.ConnectivityState;
 import io.grpc.ManagedChannel;
+import io.grpc.MethodDescriptor;
 
 /**
  * Exposes Flight GRPC service & client.
  */
 public class FlightGrpcUtils {
+  /**
+   * Proxy class for ManagedChannel that makes closure a no-op.
+   */
+  static class ProxyManagedChannel extends ManagedChannel {
+    private final ManagedChannel channel;
+    private boolean isShutdown;
+    private final List<Runnable> shutdownCallbacks;
+
+    ProxyManagedChannel(ManagedChannel channel) {
+      this.channel = channel;
+      this.isShutdown = channel.isShutdown();
+      this.shutdownCallbacks = new ArrayList<>();
+    }
+
+    @Override
+    public ManagedChannel shutdown() {
+      isShutdown = true;
+      shutdownCallbacks.forEach(c -> c.run());
+      return this;
+    }
+
+    @Override
+    public boolean isShutdown() {
+      if (this.channel.isShutdown()) {
+        // If the underlying channel is shut down, ensure we're updated to match.
+        shutdown();
+      }
+      return isShutdown;
+    }
+
+    @Override
+    public boolean isTerminated() {
+      return this.isShutdown();
+    }
+
+    @Override
+    public ManagedChannel shutdownNow() {
+      return shutdown();
+    }
+
+    @Override
+    public boolean awaitTermination(long l, TimeUnit timeUnit) {
+      // Don't actually await termination, since it'll be a no-op, so simply return whether or not
+      // the channel has been shut down already.
+      return this.isShutdown();
+    }
+
+    @Override
+    public <RequestT, ResponseT> ClientCall<RequestT, ResponseT> newCall(
+        MethodDescriptor<RequestT, ResponseT> methodDescriptor, CallOptions callOptions) {
+      if (this.isShutdown()) {
+        throw new IllegalStateException("Channel has been shut down.");
+      }
+
+      return this.channel.newCall(methodDescriptor, callOptions);
+    }
+
+    @Override
+    public String authority() {
+      return this.channel.authority();
+    }
+
+    @Override
+    public ConnectivityState getState(boolean requestConnection) {
+      if (this.isShutdown()) {
+        return ConnectivityState.SHUTDOWN;
+      }
+
+      return this.channel.getState(requestConnection);
+    }
+
+    @Override
+    public void notifyWhenStateChanged(ConnectivityState source, Runnable callback) {
+      if (source == ConnectivityState.SHUTDOWN) {
```

Comment:
Wait, this is wrong. This is the current state before the state change, not the new state.

Reply:
This is supposed to also run in the same executor that runs call listeners.

Reply:
Upon trying to think through this, I don't think it makes sense to track the state changes here. This would entail essentially recreating the entire ManagedChannel machine for tracking state, which is not really feasible.